// @wordpress/block-editor/build-module/components/block-styles/preview-panel.mjs
// packages/block-editor/src/components/block-styles/preview-panel.js
import { useMemo } from "@wordpress/element";
import InserterPreviewPanel from "../inserter/preview-panel.mjs";
import { replaceActiveStyle } from "./utils.mjs";
import { jsx } from "react/jsx-runtime";
function BlockStylesPreviewPanel({
  genericPreviewBlock,
  style,
  className,
  activeStyle
}) {
  const styleClassName = replaceActiveStyle(className, activeStyle, style);
  const previewBlocks = useMemo(() => {
    return {
      name: genericPreviewBlock.name,
      title: style.label || style.name,
      description: style.description,
      initialAttributes: {
        ...genericPreviewBlock.attributes,
        className: styleClassName + " block-editor-block-styles__block-preview-container"
      },
      example: genericPreviewBlock
    };
  }, [genericPreviewBlock, style, styleClassName]);
  return /* @__PURE__ */ jsx(InserterPreviewPanel, { item: previewBlocks });
}
export {
  BlockStylesPreviewPanel as default
};
