// @wordpress/upload-media/build/store/private-actions.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/upload-media/src/store/private-actions.ts
var private_actions_exports = {};
__export(private_actions_exports, {
  addItem: () => addItem,
  finishOperation: () => finishOperation,
  pauseItem: () => pauseItem,
  pauseQueue: () => pauseQueue,
  prepareItem: () => prepareItem,
  processItem: () => processItem,
  removeItem: () => removeItem,
  resumeItem: () => resumeItem,
  resumeQueue: () => resumeQueue,
  revokeBlobUrls: () => revokeBlobUrls,
  updateItemProgress: () => updateItemProgress,
  updateSettings: () => updateSettings,
  uploadItem: () => uploadItem
});
module.exports = __toCommonJS(private_actions_exports);
var import_uuid = require("uuid");
var import_blob = require("@wordpress/blob");
var import_utils = require("../utils.cjs");
var import_stub_file = require("../stub-file.cjs");
var import_types = require("./types.cjs");
function addItem({
  file: fileOrBlob,
  batchId,
  onChange,
  onSuccess,
  onBatchSuccess,
  onError,
  additionalData = {},
  sourceUrl,
  sourceAttachmentId,
  abortController,
  operations
}) {
  return async ({ dispatch }) => {
    const itemId = (0, import_uuid.v4)();
    const file = (0, import_utils.convertBlobToFile)(fileOrBlob);
    let blobUrl;
    if (!(file instanceof import_stub_file.StubFile)) {
      blobUrl = (0, import_blob.createBlobURL)(file);
      dispatch({
        type: import_types.Type.CacheBlobUrl,
        id: itemId,
        blobUrl
      });
    }
    dispatch({
      type: import_types.Type.Add,
      item: {
        id: itemId,
        batchId,
        status: import_types.ItemStatus.Processing,
        sourceFile: (0, import_utils.cloneFile)(file),
        file,
        attachment: {
          url: blobUrl
        },
        additionalData: {
          convert_format: false,
          ...additionalData
        },
        onChange,
        onSuccess,
        onBatchSuccess,
        onError,
        sourceUrl,
        sourceAttachmentId,
        abortController: abortController || new AbortController(),
        operations: Array.isArray(operations) ? operations : [import_types.OperationType.Prepare]
      }
    });
    dispatch.processItem(itemId);
  };
}
function processItem(id) {
  return async ({ select, dispatch }) => {
    if (select.isPaused()) {
      return;
    }
    const item = select.getItem(id);
    const { attachment, onChange, onSuccess, onBatchSuccess, batchId } = item;
    const operation = Array.isArray(item.operations?.[0]) ? item.operations[0][0] : item.operations?.[0];
    if (operation === import_types.OperationType.Upload) {
      const settings = select.getSettings();
      const activeCount = select.getActiveUploadCount();
      if (activeCount >= settings.maxConcurrentUploads) {
        return;
      }
    }
    if (attachment) {
      onChange?.([attachment]);
    }
    if (!operation) {
      if (attachment) {
        onSuccess?.([attachment]);
      }
      dispatch.revokeBlobUrls(id);
      if (batchId && select.isBatchUploaded(batchId)) {
        onBatchSuccess?.();
      }
      return;
    }
    if (!operation) {
      return;
    }
    dispatch({
      type: import_types.Type.OperationStart,
      id,
      operation
    });
    switch (operation) {
      case import_types.OperationType.Prepare:
        dispatch.prepareItem(item.id);
        break;
      case import_types.OperationType.Upload:
        dispatch.uploadItem(id);
        break;
    }
  };
}
function pauseQueue() {
  return {
    type: import_types.Type.PauseQueue
  };
}
function resumeQueue() {
  return async ({ select, dispatch }) => {
    dispatch({
      type: import_types.Type.ResumeQueue
    });
    for (const item of select.getAllItems()) {
      dispatch.processItem(item.id);
    }
  };
}
function pauseItem(id) {
  return async ({ dispatch }) => {
    dispatch({
      type: import_types.Type.PauseItem,
      id
    });
  };
}
function resumeItem(id) {
  return async ({ select, dispatch }) => {
    const item = select.getItem(id);
    if (!item || item.status !== import_types.ItemStatus.Paused) {
      return;
    }
    dispatch({
      type: import_types.Type.ResumeItem,
      id
    });
    dispatch.processItem(id);
  };
}
function removeItem(id) {
  return async ({ select, dispatch }) => {
    const item = select.getItem(id);
    if (!item) {
      return;
    }
    dispatch({
      type: import_types.Type.Remove,
      id
    });
  };
}
function finishOperation(id, updates) {
  return async ({ select, dispatch }) => {
    const item = select.getItem(id);
    const previousOperation = item?.currentOperation;
    dispatch({
      type: import_types.Type.OperationFinish,
      id,
      item: updates
    });
    dispatch.processItem(id);
    if (previousOperation === import_types.OperationType.Upload) {
      const pendingUploads = select.getPendingUploads();
      for (const pendingItem of pendingUploads) {
        dispatch.processItem(pendingItem.id);
      }
    }
  };
}
function prepareItem(id) {
  return async ({ dispatch }) => {
    const operations = [import_types.OperationType.Upload];
    dispatch({
      type: import_types.Type.AddOperations,
      id,
      operations
    });
    dispatch.finishOperation(id, {});
  };
}
function uploadItem(id) {
  return async ({ select, dispatch }) => {
    const item = select.getItem(id);
    select.getSettings().mediaUpload({
      filesList: [item.file],
      additionalData: item.additionalData,
      signal: item.abortController?.signal,
      onFileChange: ([attachment]) => {
        if (!(0, import_blob.isBlobURL)(attachment.url)) {
          dispatch.finishOperation(id, {
            attachment
          });
        }
      },
      onSuccess: ([attachment]) => {
        dispatch.finishOperation(id, {
          attachment
        });
      },
      onError: (error) => {
        dispatch.cancelItem(id, error);
      }
    });
  };
}
function revokeBlobUrls(id) {
  return async ({ select, dispatch }) => {
    const blobUrls = select.getBlobUrls(id);
    for (const blobUrl of blobUrls) {
      (0, import_blob.revokeBlobURL)(blobUrl);
    }
    dispatch({
      type: import_types.Type.RevokeBlobUrls,
      id
    });
  };
}
function updateItemProgress(id, progress) {
  return async ({ dispatch }) => {
    dispatch({
      type: import_types.Type.UpdateProgress,
      id,
      progress
    });
  };
}
function updateSettings(settings) {
  return {
    type: import_types.Type.UpdateSettings,
    settings
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  addItem,
  finishOperation,
  pauseItem,
  pauseQueue,
  prepareItem,
  processItem,
  removeItem,
  resumeItem,
  resumeQueue,
  revokeBlobUrls,
  updateItemProgress,
  updateSettings,
  uploadItem
});
