// @wordpress/block-editor/build-module/store/private-keys.mjs
// packages/block-editor/src/store/private-keys.js
var globalStylesDataKey = /* @__PURE__ */ Symbol("globalStylesDataKey");
var globalStylesLinksDataKey = /* @__PURE__ */ Symbol("globalStylesLinks");
var selectBlockPatternsKey = /* @__PURE__ */ Symbol("selectBlockPatternsKey");
var reusableBlocksSelectKey = /* @__PURE__ */ Symbol("reusableBlocksSelect");
var sectionRootClientIdKey = /* @__PURE__ */ Symbol("sectionRootClientIdKey");
var mediaEditKey = /* @__PURE__ */ Symbol("mediaEditKey");
var getMediaSelectKey = /* @__PURE__ */ Symbol("getMediaSelect");
var essentialFormatKey = /* @__PURE__ */ Symbol("essentialFormat");
var isIsolatedEditorKey = /* @__PURE__ */ Symbol("isIsolatedEditor");
var deviceTypeKey = /* @__PURE__ */ Symbol("deviceTypeKey");
export {
  deviceTypeKey,
  essentialFormatKey,
  getMediaSelectKey,
  globalStylesDataKey,
  globalStylesLinksDataKey,
  isIsolatedEditorKey,
  mediaEditKey,
  reusableBlocksSelectKey,
  sectionRootClientIdKey,
  selectBlockPatternsKey
};
