// @wordpress/block-editor/build-module/index.mjs
// packages/block-editor/src/index.js
import "./hooks/index.mjs";
import {
  getBorderClassesAndStyles,
  useBorderProps,
  getColorClassesAndStyles,
  getTypographyClassesAndStyles,
  useColorProps,
  useCustomSides,
  getSpacingClassesAndStyles,
  getGapCSSValue,
  getShadowClassesAndStyles,
  useCachedTruthy,
  useStyleOverride
} from "./hooks/index.mjs";
export * from "./components/index.mjs";
export * from "./elements/index.mjs";
export * from "./utils/index.mjs";
import { storeConfig, store } from "./store/index.mjs";
import { SETTINGS_DEFAULTS } from "./store/defaults.mjs";
import { privateApis } from "./private-apis.mjs";
export {
  SETTINGS_DEFAULTS,
  getBorderClassesAndStyles as __experimentalGetBorderClassesAndStyles,
  getColorClassesAndStyles as __experimentalGetColorClassesAndStyles,
  getGapCSSValue as __experimentalGetGapCSSValue,
  getShadowClassesAndStyles as __experimentalGetShadowClassesAndStyles,
  getSpacingClassesAndStyles as __experimentalGetSpacingClassesAndStyles,
  useBorderProps as __experimentalUseBorderProps,
  useColorProps as __experimentalUseColorProps,
  useCustomSides as __experimentalUseCustomSides,
  getTypographyClassesAndStyles,
  privateApis,
  store,
  storeConfig,
  useCachedTruthy,
  useStyleOverride
};
