// @wordpress/block-editor/build/components/block-visibility/viewport-menu-item.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/block-editor/src/components/block-visibility/viewport-menu-item.js
var viewport_menu_item_exports = {};
__export(viewport_menu_item_exports, {
  default: () => BlockVisibilityViewportMenuItem
});
module.exports = __toCommonJS(viewport_menu_item_exports);
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_icons = require("@wordpress/icons");
var import_element = require("@wordpress/element");
var import_data = require("@wordpress/data");
var import__ = require("./index.cjs");
var import_store = require("../../store/index.cjs");
var import_lock_unlock = require("../../lock-unlock.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function BlockVisibilityViewportMenuItem({ clientIds }) {
  const [isModalOpen, setIsModalOpen] = (0, import_element.useState)(false);
  const { areBlocksHiddenAnywhere } = (0, import_data.useSelect)(
    (select) => {
      const { isBlockHiddenAnywhere } = (0, import_lock_unlock.unlock)(
        select(import_store.store)
      );
      return {
        areBlocksHiddenAnywhere: clientIds?.every(
          (clientId) => isBlockHiddenAnywhere(clientId)
        )
      };
    },
    [clientIds]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_components.MenuItem,
      {
        icon: areBlocksHiddenAnywhere ? import_icons.unseen : import_icons.seen,
        onClick: () => setIsModalOpen(true),
        children: areBlocksHiddenAnywhere ? (0, import_i18n.__)("Show") : (0, import_i18n.__)("Hide")
      }
    ),
    isModalOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import__.BlockVisibilityModal,
      {
        clientIds,
        onClose: () => setIsModalOpen(false)
      }
    )
  ] });
}
